x-pack/plugins/alerting/server/alerts_client/alerts_client.ts

```diff
@@ -0,0 +1,295 @@
+        : DEFAULT_NAMESPACE_STRING,
+    });
+
+    const { alertsToReturn, recoveredAlertsToReturn } =
+      await this.legacyAlertsClient.getAlertsToSerialize(false);
+
+    const activeAlerts = this.legacyAlertsClient.getProcessedAlerts('active');
+    const recoveredAlerts = this.legacyAlertsClient.getProcessedAlerts('recovered');
+
+    // TODO - Lifecycle alerts set some other fields based on alert status
+    // Example: workflow status - default to 'open' if not set
+    // event action: new alert = 'new', active alert: 'active', otherwise 'close'
+
+    const activeAlertsToIndex: Array<Alert & AlertData> = [];
+    for (const id of keys(alertsToReturn)) {
+      // See if there's an existing active alert document
+      if (
+        this.fetchedAlerts.data.hasOwnProperty(id) &&
+        this.fetchedAlerts.data[id].kibana.alert.status === 'active'
+      ) {
+        activeAlertsToIndex.push(
+          buildOngoingAlert<
+            AlertData,
+            LegacyState,
+            LegacyContext,
+            ActionGroupIds,
+            RecoveryActionGroupId
+          >({
+            alert: this.fetchedAlerts.data[id],
+            legacyAlert: activeAlerts[id],
+            rule: this.rule,
+            timestamp: currentTime,
+          })
+        );
+      } else {
+        activeAlertsToIndex.push(
+          buildNewAlert<
+            AlertData,
+            LegacyState,
+            LegacyContext,
+            ActionGroupIds,
+            RecoveryActionGroupId
+          >({ legacyAlert: activeAlerts[id], rule: this.rule, timestamp: currentTime })
+        );
+      }
+    }
+
+    const recoveredAlertsToIndex: Array<Alert & AlertData> = [];
+    for (const id of keys(recoveredAlertsToReturn)) {
+      // See if there's an existing alert document
+      // If there is not, log an error because there should be
+      if (this.fetchedAlerts.data.hasOwnProperty(id)) {
+        recoveredAlertsToIndex.push(
+          buildRecoveredAlert<
+            AlertData,
+            LegacyState,
+            LegacyContext,
+            ActionGroupIds,
+            RecoveryActionGroupId
+          >({
+            alert: this.fetchedAlerts.data[id],
+            legacyAlert: recoveredAlerts[id],
+            rule: this.rule,
+            timestamp: currentTime,
+          })
+        );
+      } else {
+        this.options.logger.warn(
+          `Could not find alert document to update for recovered alert with id ${id} and uuid ${recoveredAlerts[
+            id
+          ].getUuid()}`
+        );
+      }
+    }
+
+    const alertsToIndex = [...activeAlertsToIndex, ...recoveredAlertsToIndex];
+    if (alertsToIndex.length > 0) {
+      await esClient.bulk({
+        refresh: 'wait_for',
+        index: indexTemplateAndPattern.alias,
+        require_alias: true,
+        body: flatMap(
+          [...activeAlertsToIndex, ...recoveredAlertsToIndex].map((alert: Alert & AlertData) => [
+            {
+              index: {
```

Comment:
**question:** What happens if the bulk call happens at the same time a user updates the same alert document? Will this overwrite the user's action and/or vice versa?

Reply:
That's a good question! I think that's something that could happen currently with the existing flow right? I think we could create a followup issue to discuss the correct behavior (like which update gets priority) unless you feel like it's something necessary to address in this PR?

Reply:
Issue created here: https://github.com/elastic/kibana/issues/158403